models/repo.go

```diff
@@ -2452,3 +2471,101 @@ func (repo *Repository) GetUserFork(userID int64) (*Repository, error) {
 	}
 	return &forkedRepo, nil
 }
+
+// CustomAvatarPath returns repository custom avatar file path.
+func (repo *Repository) CustomAvatarPath() string {
+	// Avatar empty by default
+	if len(repo.Avatar) <= 0 {
+		return ""
+	}
+	return filepath.Join(setting.RepositoryAvatarUploadPath, repo.Avatar)
+}
+
+// RelAvatarLink returns a relative link to the user's avatar.
+// The link a sub-URL to this site
+// Since Gravatar support not needed here - just check for image path.
+func (repo *Repository) RelAvatarLink() string {
+	// If no avatar - path is empty
+	avatarPath := repo.CustomAvatarPath()
+	if len(avatarPath) <= 0 {
+		return ""
+	}
+	if !com.IsFile(avatarPath) {
+		return ""
+	}
+	return setting.AppSubURL + "/repo-avatars/" + repo.Avatar
+}
+
+// AvatarLink returns user avatar absolute link.
+func (repo *Repository) AvatarLink() string {
+	link := repo.RelAvatarLink()
+	// link may be empty!
+	if len(link) > 0 {
+		if link[0] == '/' && link[1] != '/' {
+			return setting.AppURL + strings.TrimPrefix(link, setting.AppSubURL)[1:]
+		}
+	}
+	return link
+}
+
+// UploadAvatar saves custom avatar for repository.
+// FIXME: split uploads to different subdirs in case we have massive number of repos.
+func (repo *Repository) UploadAvatar(data []byte) error {
+	m, err := avatar.Prepare(data)
+	if err != nil {
+		return err
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	oldAvatarPath := repo.CustomAvatarPath()
+
+	// Users can upload the same image to other repo - prefix it with ID
+	// Then repo will be removed - only it avatar file will be removed
+	repo.Avatar = fmt.Sprintf("%d-%x", repo.ID, md5.Sum(data))
+	if _, err := x.ID(repo.ID).Cols("avatar").Update(repo); err != nil {
```

Comment:
`sess.` not `x.`

Reply:
Done in https://github.com/go-gitea/gitea/pull/6986/commits/c007a9a947343f4a11589177e3872f51a709989d